File: 1.300/205.IsomorphicStrings.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class IsomorphicStrings
{
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size())
            return false;
        unordered_map<char, char> dp;
        unordered_set<char> ep;
        for (int i = 0;i < s.size(); ++i) {
            auto it = dp.find(s[i]);
            if (it == dp.end()) {
                if (ep.find(t[i]) != ep.end())
                    return false;
                dp[s[i]] = t[i];
                ep.insert(t[i]);
                continue;
            }
            if (it->second != t[i])
                return false;
        }
        return true;
    }
};

int main(int argc, char const* argv[])
{
    IsomorphicStrings cls;
    cout << boolalpha << cls.isIsomorphic("badc", "baba") << " -> false" << endl;
    cout << boolalpha << cls.isIsomorphic("egg", "add") << " -> true" << endl;
    cout << boolalpha << cls.isIsomorphic("foo", "bar") << " -> false" << endl;
    cout << boolalpha << cls.isIsomorphic("paper", "title") << " -> true" << endl;
    return 0;
}

File: go/1.300/205.IsomorphicStrings.go
package main

import "fmt"

func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	size := len(s)
	dp := map[byte]byte{}
	ep := map[byte]byte{}
	for i := 0; i < size; i++ {
		if c, ok := dp[s[i]]; ok {
			if c != t[i] {
				return false
			} else {
				continue
			}
		}
		if _, ok := ep[t[i]]; ok {
			return false
		}
		dp[s[i]] = t[i]
		ep[t[i]] = s[i]
	}
	return true
}

func main() {
	fmt.Printf("%v -> false \n", isIsomorphic("badc", "baba"))
	fmt.Printf("%v -> true \n", isIsomorphic("egg", "add"))
	fmt.Printf("%v -> false \n", isIsomorphic("foo", "bar"))
	fmt.Printf("%v -> true \n", isIsomorphic("paper", "title"))
}
